File: go/benchmark.go
package main

import (
	"os"
	"math"
	"fmt"

	"github.com/sjwhitworth/golearn/base"
	"github.com/sjwhitworth/golearn/ensemble"
	"github.com/sjwhitworth/golearn/evaluation"

	"github.com/Anupam-Hari/cuml-go/go/internal/capi"
	datasetpkg "github.com/Anupam-Hari/cuml-go/go/internal/dataset"
	"github.com/Anupam-Hari/cuml-go/go/kmeans"
	"github.com/Anupam-Hari/cuml-go/go/knn"
	"github.com/Anupam-Hari/cuml-go/go/random_forest"
)

func BenchmarkRandomForest(dataset Dataset) (BenchmarkResult, error) {

	split := SplitDataset(dataset, 0.8)

	result := BenchmarkResult{
		Model:     "Random Forest",
		TrainRows: split.TrainRows,
		TestRows:  split.TestRows,
	}

	rf, err := randomforest.New(
		randomforest.WithEstimators(100),
		randomforest.WithMaxDepth(10),
		randomforest.WithMaxFeatures(1.0),
		randomforest.WithMaxLeaves(-1),
		randomforest.WithMaxSamples(1.0),
	)
	if err != nil {
		return result, err
	}
	defer rf.Close()

	cpuMonitor, _ := capi.NewCPUMonitor()
	defer cpuMonitor.Close()

	gpumonitor, _ := capi.NewGPUMonitor()
	defer gpumonitor.Close()

	timer := Timer{}

	gpumonitor.Start()
	cpuMonitor.Start()

	timer.Start()

	err = rf.Fit(
		split.XTrain,
		split.YTrain,
	)
	if err != nil {
		return result, err
	}

	result.TrainTimeMS = timer.Stop()

	timer.Start()

	pred, err := rf.Predict(split.XTest, 1)
	if err != nil {
		return result, err
	}

	predictTimeMS := timer.Stop()
	gpumonitor.Stop()
	cpuMonitor.Stop()

	correct := 0
	for i := range pred {
		if pred[i] == split.YTest[i] {
			correct++
		}
	}

	accuracy := float64(correct) / float64(len(split.YTest))
	result.Accuracy = accuracy

	result.GPUAvg = gpumonitor.Average()
	result.GPUPeak = gpumonitor.Peak()

	result.CPUAvg = cpuMonitor.Average()
	result.CPUPeak = cpuMonitor.Peak()

	result.PredictionThroughput =
		float64(result.TestRows) /
			(predictTimeMS / 1000.0)

	result.TotalTimeMS =
		result.TrainTimeMS +
			predictTimeMS

	return result, nil
}

func BenchmarkGoLearnRandomForest(dataset Dataset) (BenchmarkResult, error) {

	split := SplitDataset(dataset, 0.8)

	result := BenchmarkResult{
		Model:     "GoLearn Random Forest",
		TrainRows: split.TrainRows,
		TestRows:  split.TestRows,
	}

	trainFile := "golearn_train_tmp.csv"
	testFile := "golearn_test_tmp.csv"

	err := datasetpkg.WriteGoLearnCSV(
		trainFile,
		split.XTrain,
		split.YTrain,
	)
	if err != nil {
		return result, err
	}

	err = datasetpkg.WriteGoLearnCSV(
		testFile,
		split.XTest,
		split.YTest,
	)
	if err != nil {
		return result, err
	}

	defer os.Remove(trainFile)
	defer os.Remove(testFile)

	classAttr := base.NewCategoricalAttribute()
	classAttr.SetName("class")

	overrides := map[int]base.Attribute{
		49: classAttr,
	}

	classGroups := map[string]string{
		"class": "CLASS",
	}

	trainData, err := base.ParseCSVToInstancesWithAttributeGroups(
		trainFile,
		nil,
		classGroups,
		overrides,
		true,
	)
	if err != nil {
		return result, err
	}

	testData, err := base.ParseCSVToInstancesWithAttributeGroups(
		testFile,
		nil,
		classGroups,
		overrides,
		true,
	)
	if err != nil {
		return result, err
	}

	mtry := int(math.Sqrt(float64(len(split.XTrain[0]))))

	rf := ensemble.NewRandomForest(
		100, // trees
		mtry,   // features considered per split
	)
	cpuMonitor, _ := capi.NewCPUMonitor()
	defer cpuMonitor.Close()

	timer := Timer{}
	cpuMonitor.Start()

	timer.Start()

	err = rf.Fit(trainData)
	if err != nil {
		return result, err
	}

	result.TrainTimeMS = timer.Stop()

	timer.Start()

	predictions, err := rf.Predict(testData)
	if err != nil {
		return result, err
	}

	predictTimeMS := timer.Stop()

	cpuMonitor.Stop()

	cm, err := evaluation.GetConfusionMatrix(
		testData,
		predictions,
	)
	if err != nil {
		return result, err
	}

	result.CPUAvg = cpuMonitor.Average()
	result.CPUPeak = cpuMonitor.Peak()

	result.Accuracy = evaluation.GetAccuracy(cm)

	result.PredictionThroughput =
		float64(result.TestRows) /
		(predictTimeMS / 1000.0)

	result.TotalTimeMS =
		result.TrainTimeMS +
		predictTimeMS

	return result, nil
}

func BenchmarkRFInference(dataset Dataset) (BenchmarkResult, error) {

	split := SplitDataset(dataset, 0.8)

	result := BenchmarkResult{
		Model:     "Random Forest Inference",
		TrainRows: split.TrainRows,
		TestRows:  split.TestRows,
	}

	rf, err := randomforest.New(
		randomforest.WithEstimators(100),
		randomforest.WithMaxDepth(10),
		randomforest.WithMaxFeatures(1.0),
		randomforest.WithMaxLeaves(-1),
		randomforest.WithMaxSamples(1.0),
	)
	if err != nil {
		return result, err
	}
	defer rf.Close()

	timer := Timer{}

	//-------------------------------------------------
	// Train once
	//-------------------------------------------------

	timer.Start()

	err = rf.Fit(
		split.XTrain,
		split.YTrain,
	)
	if err != nil {
		return result, err
	}

	result.TrainTimeMS = timer.Stop()

	//-------------------------------------------------
	// GPU warmup
	//-------------------------------------------------

	_, err = rf.Predict(
		split.XTest,
		randomforest.BackendGPU,
	)
	if err != nil {
		return result, err
	}

	//-------------------------------------------------
	// GPU benchmark
	//-------------------------------------------------

	timer.Start()

	gpuPred, err := rf.Predict(
		split.XTest,
		randomforest.BackendGPU,
	)
	if err != nil {
		return result, err
	}

	result.GPUPredictionTimeMS = timer.Stop()

	result.GPUThroughput =
		float64(result.TestRows) /
			(result.GPUPredictionTimeMS / 1000.0)

	//-------------------------------------------------
	// CPU warmup
	//-------------------------------------------------

	_, err = rf.Predict(
		split.XTest,
		randomforest.BackendCPU,
	)
	if err != nil {
		return result, err
	}

	//-------------------------------------------------
	// CPU benchmark
	//-------------------------------------------------

	timer.Start()

	cpuPred, err := rf.Predict(
		split.XTest,
		randomforest.BackendCPU,
	)
	if err != nil {
		return result, err
	}

	result.CPUPredictionTimeMS = timer.Stop()

	result.CPUThroughput =
		float64(result.TestRows) /
			(result.CPUPredictionTimeMS / 1000.0)

	//-------------------------------------------------
	// Accuracy
	//-------------------------------------------------

	correct := 0

	for i := range gpuPred {

		if gpuPred[i] == split.YTest[i] {
			correct++
		}

		if gpuPred[i] != cpuPred[i] {
			return result, fmt.Errorf(
				"CPU and GPU predictions differ at sample %d",
				i,
			)
		}
	}

	result.Accuracy =
		float64(correct) /
			float64(len(split.YTest))

	return result, nil
}

func BenchmarkKNN(dataset Dataset) (BenchmarkResult, error) {

	split := SplitDataset(dataset, 0.8)

	result := BenchmarkResult{
		Model:     "KNN",
		TrainRows: split.TrainRows,
		TestRows:  split.TestRows,
	}

	knn, err := knn.New(
		knn.WithK(5),
	)
	if err != nil {
		return result, err
	}
	defer knn.Close()

	timer := Timer{}
	gpumonitor, _ := capi.NewGPUMonitor()
	defer gpumonitor.Close()

	cpuMonitor, _ := capi.NewCPUMonitor()
    defer cpuMonitor.Close()

	gpumonitor.Start()
	cpuMonitor.Start()
	timer.Start()

	err = knn.Fit(
		split.XTrain,
		split.YTrain,
	)
	if err != nil {
		return result, err
	}

	result.TrainTimeMS = timer.Stop()

	timer.Start()

	pred, err := knn.Predict(split.XTest)
	if err != nil {
		return result, err
	}

	predictTimeMS := timer.Stop()

	gpumonitor.Stop()
	cpuMonitor.Stop()

	correct := 0
	for i := range pred {
		if pred[i] == split.YTest[i] {
			correct++
		}
	}

	accuracy := float64(correct) / float64(len(split.YTest))

	result.Accuracy = accuracy

	result.GPUAvg = gpumonitor.Average()
	result.GPUPeak = gpumonitor.Peak()

	result.CPUAvg = cpuMonitor.Average()
	result.CPUPeak = cpuMonitor.Peak()

	result.PredictionThroughput =
		float64(result.TestRows) /
			(predictTimeMS / 1000.0)

	result.TotalTimeMS =
		result.TrainTimeMS +
			predictTimeMS

	return result, nil
}

func BenchmarkKMeans(dataset Dataset) (BenchmarkResult, error) {

	split := SplitDataset(dataset, 0.8)

	result := BenchmarkResult{
		Model:     "KMeans",
		TrainRows: split.TrainRows,
		TestRows:  split.TestRows,
	}

	nClasses, err := NumClasses(dataset)
	if err != nil {
		return result, err
	}

	kmeans, err := kmeans.New(
		kmeans.WithNClusters(nClasses),
		kmeans.WithMaxIters(300),
		kmeans.WithTolerance(1e-4),
	)
	if err != nil {
		return result, err
	}
	defer kmeans.Close()

	timer := Timer{}

	cpuMonitor, _ := capi.NewCPUMonitor()
	defer cpuMonitor.Close()

	gpumonitor, _ := capi.NewGPUMonitor()
	defer gpumonitor.Close()

	cpuMonitor.Start()
	gpumonitor.Start()
	

	timer.Start()

	err = kmeans.Fit(split.XTrain)
	if err != nil {
		return result, err
	}

	result.TrainTimeMS = timer.Stop()

	timer.Start()

	_, err = kmeans.Predict(split.XTest)
	if err != nil {
		return result, err
	}

	predictTimeMS := timer.Stop()

	gpumonitor.Stop()
	cpuMonitor.Stop()

	result.GPUAvg = gpumonitor.Average()
	result.GPUPeak = gpumonitor.Peak()

	result.CPUAvg = cpuMonitor.Average()
	result.CPUPeak = cpuMonitor.Peak()

	result.PredictionThroughput =
		float64(result.TestRows) /
			(predictTimeMS / 1000.0)

	result.TotalTimeMS =
		result.TrainTimeMS +
			predictTimeMS

	return result, nil
}

File: benchmark/benchmark.cpp
#include "benchmark.h"

#include "timer.h"

#include "../cpp/random_forest/random_forest.h"
#include "../cpp/knn/knn.h"
#include "../cpp/kmeans/kmeans.h"

#include <algorithm>
#include <vector>

namespace
{

struct TrainTestSplit
{
    std::vector<float> X_train;
    std::vector<int> y_train;

    std::vector<float> X_test;
    std::vector<int> y_test;

    int train_rows = 0;
    int test_rows = 0;
};

TrainTestSplit split_dataset(
    const Dataset& dataset,
    float train_ratio = 0.8f)
{
    TrainTestSplit split;

    split.train_rows =
        static_cast<int>(dataset.rows * train_ratio);

    split.test_rows =
        dataset.rows - split.train_rows;

    split.X_train.reserve(split.train_rows * dataset.cols);
    split.y_train.reserve(split.train_rows);

    split.X_test.reserve(split.test_rows * dataset.cols);
    split.y_test.reserve(split.test_rows);

    for (int i = 0; i < dataset.rows; i++)
    {
        const float* row =
            dataset.X.data() + i * dataset.cols;

        if (i < split.train_rows)
        {
            split.X_train.insert(
                split.X_train.end(),
                row,
                row + dataset.cols);

            split.y_train.push_back(dataset.y[i]);
        }
        else
        {
            split.X_test.insert(
                split.X_test.end(),
                row,
                row + dataset.cols);

            split.y_test.push_back(dataset.y[i]);
        }
    }

    return split;
}

int num_classes(const Dataset& dataset)
{
    int max_class = *std::max_element(
        dataset.y.begin(),
        dataset.y.end());

    return max_class + 1;
}

}

BenchmarkResult benchmark_random_forest(
    const Dataset& dataset)
{
    auto split = split_dataset(dataset);

    BenchmarkResult result;

    result.model = "Random Forest";
    result.train_rows = split.train_rows;
    result.test_rows = split.test_rows;

    RFHandle* rf = rf_create(
        100,
        10,
        1.0f,
        -1,
        1.0f);

    Timer timer;

    timer.start();

    rf_fit(
        rf,
        split.X_train.data(),
        split.train_rows,
        dataset.cols,
        split.y_train.data(),
        num_classes(dataset));

    result.train_time_ms = timer.stop();

    std::vector<int> predictions(split.test_rows);

    timer.start();

    rf_predict(
        rf,
        split.X_test.data(),
        split.test_rows,
        dataset.cols,
        predictions.data());

    result.predict_time_ms = timer.stop();

    result.total_time_ms =
        result.train_time_ms +
        result.predict_time_ms;

    rf_destroy(rf);

    return result;
}

BenchmarkResult benchmark_knn(
    const Dataset& dataset)
{
    auto split = split_dataset(dataset);

    BenchmarkResult result;

    result.model = "KNN";
    result.train_rows = split.train_rows;
    result.test_rows = split.test_rows;

    KNNHandle* knn = knn_create(5);

    Timer timer;

    timer.start();

    knn_fit(
        knn,
        split.X_train.data(),
        split.train_rows,
        dataset.cols,
        split.y_train.data());

    result.train_time_ms = timer.stop();

    std::vector<int> predictions(split.test_rows);

    timer.start();

    knn_predict(
        knn,
        split.X_test.data(),
        split.test_rows,
        dataset.cols,
        predictions.data());

    result.predict_time_ms = timer.stop();

    result.total_time_ms =
        result.train_time_ms +
        result.predict_time_ms;

    knn_destroy(knn);

    return result;
}

BenchmarkResult benchmark_kmeans(
    const Dataset& dataset)
{
    auto split = split_dataset(dataset);

    BenchmarkResult result;

    result.model = "KMeans";
    result.train_rows = split.train_rows;
    result.test_rows = split.test_rows;

    KMeansHandle* km = kmeans_create(
        num_classes(dataset),
        8,
        1e-4f);

    Timer timer;

    timer.start();

    kmeans_fit(
        km,
        split.X_train.data(),
        split.train_rows,
        dataset.cols);

    result.train_time_ms = timer.stop();

    std::vector<int> labels(split.test_rows);

    timer.start();

    kmeans_predict(
        km,
        split.X_test.data(),
        split.test_rows,
        dataset.cols,
        labels.data());

    result.predict_time_ms = timer.stop();

    result.total_time_ms =
        result.train_time_ms +
        result.predict_time_ms;

    kmeans_destroy(km);

    return result;
}